PWGJE/EMCALJetTasks/UserTasks/AliAnalysisTaskEmcalVsPhos.cxx

```diff
@@ -685,6 +691,124 @@ void AliAnalysisTaskEmcalVsPhos::AllocateClustersInJetsHistograms()
   }
 }
 
+/*
+ * This function allocates the histograms for the jet performance study.
+ */
+void AliAnalysisTaskEmcalVsPhos::AllocateJetPerformanceHistograms()
+{
+  TString histname;
+  TString htitle;
+  Int_t nPtBins = TMath::CeilNint(fMaxPt/2);
+  
+  const Int_t nParticleTypes = 8;
+  Double_t *particleTypeBins = GenerateFixedBinArray(nParticleTypes, -0.5, 7.5);
+  
+  // M02 vs. Energy vs. Particle type
+  histname = "JetPerformance/hM02VsParticleType";
+  htitle = histname + ";M02;#it{E}_{clus} (GeV); Particle type";
+  fHistManager.CreateTH3(histname.Data(), htitle.Data(), fNM02HistBins, fM02HistBins, fNPtHistBins, fPtHistBins, nParticleTypes, particleTypeBins);
+  
+  // M02 vs. Energy vs. Particle type vs. Jet pT, for particles inside jets
+  Int_t dim = 0;
+  TString title[20];
+  Int_t nbins[20] = {0};
+  Double_t min[30] = {0.};
+  Double_t max[30] = {0.};
+  Double_t *binEdges[20] = {0};
+  
+  title[dim] = "M02";
+  nbins[dim] = fNM02HistBins;
+  binEdges[dim] = fM02HistBins;
+  min[dim] = fM02HistBins[0];
+  max[dim] = fM02HistBins[fNM02HistBins];
+  dim++;
+  
+  title[dim] = "#it{E}_{clus} (GeV)";
+  nbins[dim] = fNPtHistBins;
+  binEdges[dim] = fPtHistBins;
+  min[dim] = fPtHistBins[0];
+  max[dim] = fPtHistBins[fNPtHistBins];
+  dim++;
+  
+  title[dim] = "Particle type";
+  nbins[dim] = nParticleTypes;
+  min[dim] = -0.5;
+  max[dim] = 7.5;
+  binEdges[dim] = particleTypeBins;
+  dim++;
+  
+  title[dim] = "#it{p}_{T,jet}^{corr}";
+  nbins[dim] = nPtBins;
+  min[dim] = 0;
+  max[dim] = fMaxPt;
+  binEdges[dim] = GenerateFixedBinArray(nbins[dim], min[dim], max[dim]);
+  dim++;
+  
+  TString thnname = "JetPerformance/hM02VsParticleTypeJets";
+  THnSparse* hn = fHistManager.CreateTHnSparse(thnname.Data(), thnname.Data(), dim, nbins, min, max);
+  for (Int_t i = 0; i < dim; i++) {
+    hn->GetAxis(i)->SetTitle(title[i]);
+    hn->SetBinEdges(i, binEdges[i]);
+  }
+  
+  // Particle composition inside each jet -- jet pT vs. particle type vs. particle number vs. particle pT sum
+  // (One entry per jet for each particle type)
+  dim = 0;
+  
+  title[dim] = "#it{p}_{T,jet}^{corr}";
+  nbins[dim] = nPtBins;
+  min[dim] = 0;
+  max[dim] = fMaxPt;
+  binEdges[dim] = GenerateFixedBinArray(nbins[dim], min[dim], max[dim]);
+  dim++;
+  
+  title[dim] = "Particle type";
+  nbins[dim] = nParticleTypes;
+  min[dim] = -0.5;
+  max[dim] = 7.5;
+  binEdges[dim] = particleTypeBins;
+  dim++;
+  
+  title[dim] = "N";
+  nbins[dim] = 30;
+  min[dim] = -0.5;
+  max[dim] = 29.5;
+  binEdges[dim] = GenerateFixedBinArray(nbins[dim], min[dim], max[dim]);
```

Comment:
@jdmulligan For your info there are helper classes in the histmanager framework which create bin edges for various types of binnings  (have a look at TLinearBinning, TCustomBinning). Could reduce your code a bit.

Reply:
Okay, thanks for the tip!